### ISSUE TITLE
Slow invalidation performance

### ISSUE THREAD

**Issue by dratchkov** [django-cache-machine/django-cache-machine#109]
I am seeing very slow (minutes per request) performance when lists are invalidated. I traced that to the recursive flush list expansion. In its present form it allows for a flush list to be hit multiple times.
The fix is to simply check if a list has already been expanded, and if it is - skip it.

I have created PR #106 . Would someone please review and pull in the changes into the main repository.

Thanks!

**Comment by timdawborn** [comment]
+1 for this PR. It looks good to me. I just hit the same issue and implemented the same fix.

**Comment by joeweiss** [comment]
+1 for this PR. Same issue here. Fix resolved the issue (> 8 min requests here).

**Comment by varunachar** [comment]
Was this hitting your gets? I'm right now seeing that gets and mgets taking minutes.. Did this fix make it into the repo? I see that this PR is open.

**Comment by timdawborn** [comment]
Yes it was, and no it's not merged in. We manually patched our installs for 12 months before giving up on django-cache-machine due to this bug.

**Comment by dratchkov** [comment]
Did you switch to something else or just no cache at all?

**Comment by timdawborn** [comment]
We removed django-cache-machine, re-analaysed performance without any caching, and added manual caching where appropriate. It took a couple of days, but it was worth it.

**Comment by varunachar** [comment]
Why would this cause gets to slow down? I understand the invalidation process would become slow, but in my case fetching from redis was taking time.

**Comment by timdawborn** [comment]
The most important part of the PR is the addition to `caching/invalidation.py` which prevents the logic from re-searching keys it has already checked. This is what was causing the huge fetch times for us — django-cache-machine was needlessly fetching the same keys over and over and over again.

**Comment by varunachar** [comment]
Ahh..  👍

**Comment by dratchkov** [comment]
Off topic question - did you see any issues with invalidation on create? I had to resort to 'whole-model' invalidation because django-cache-mahcine didn't seem to keep track of all queries an object belongs to. Did I miss anything? I am beginning to question how useful this is if it doesn't keep track of all caches and I have to blow all cache away.....

**Comment by timdawborn** [comment]
IIRC, all issues were with invalidation.

**Comment by tobiasmcnulty** [comment]
@dratchkov has a new PR for this here: https://github.com/django-cache-machine/django-cache-machine/pull/127

Others who have commented on this issue: Does this work for you and/or does validation still work as expected?
